src/gui/qgsattributeform.cpp

```diff
@@ -551,8 +551,9 @@ void QgsAttributeForm::updateValuesDependenciesDefaultValues( const int originId
           continue;
 
         QgsExpressionContext context = createExpressionContext( updatedFeature );
-        QString value = mLayer->defaultValue( eww->fieldIdx(), updatedFeature, &context ).toString();
+        const QVariant value = mLayer->defaultValue( eww->fieldIdx(), updatedFeature, &context );
         eww->setValue( value );
+        mCurrentFormFeature.setAttribute( eww->field().name(), value );
```

Comment:
For reference, this is another unreported issue, live default values were not stored into the current form feature and hence taken into account by any related functionality.